frontend/rpm/handle_sources.go

```diff
@@ -85,39 +55,35 @@ func HandleSources(wf WorkerFunc) gwclient.BuildFunc {
 }
 
 func Dalec2SourcesLLB(worker llb.State, spec *dalec.Spec, sOpt dalec.SourceOpts, opts ...llb.ConstraintsOpt) ([]llb.State, error) {
-	// Sort the map keys so that the order is consistent This shouldn't be
-	// needed when MergeOp is supported, but when it is not this will improve
-	// cache hits for callers of this function.
-	sorted := dalec.SortMapKeys(spec.Sources)
-
-	out := make([]llb.State, 0, len(spec.Sources))
-
-	for _, k := range sorted {
-		src := spec.Sources[k]
-		isDir := dalec.SourceIsDir(src)
-
-		pg := dalec.ProgressGroup("Add spec source: " + k)
-		st, err := src.AsState(k, sOpt, append(opts, pg)...)
-		if err != nil {
-			return nil, err
-		}
-
-		if isDir {
-			out = append(out, tar(st, k+".tar.gz", append(opts, pg)...))
-		} else {
-			out = append(out, st)
-		}
+	sources, err := dalec.Sources(spec, sOpt)
+	if err != nil {
+		return nil, err
 	}
+	out := make([]llb.State, 0, len(sources))
 
 	opts = append(opts, dalec.ProgressGroup("Add gomod sources"))
 	st, err := spec.GomodDeps(sOpt, worker, opts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "error adding gomod sources")
 	}
 
-	if st != nil {
-		out = append(out, tar(*st, gomodsName+".tar.gz", opts...))
+	sorted := dalec.SortMapKeys(sources)
+	for _, k := range sorted {
+		st := sources[k]
+		if dalec.SourceIsDir(spec.Sources[k]) {
+			st = st.With(sourceTar(worker, k, opts...))
+		}
+		out = append(out, st)
 	}
 
+	if st != nil {
+		out = append(out, st.With(sourceTar(worker, gomodsName, opts...)))
+	}
```

Comment:
I am assuming the order of this appending matters of this state. Because it is set before the loop and the append after the loop, is this correct? If order doesn't matter maybe, could you move the `st` value setting right before the `st` appending to the out?

Reply:
The reason I am creating it before taring up all the sources is because it can error out and I'd rather error early rather than generate a bunch of other state unnecessarily.